Audio driver for JACK

jpcima opened the pull request:
#24274

This adds ~~partial~~ support of JACK audio, ~~for now only playback~~.

JACK is a sound server for audio which uses virtual software clients and ports, it was designed for Linux professional-level audio.

I will add capture later, which must not be difficult, after I figure how to use Godot better, which I have very little practical experience with.

- There exists pseudo-devices, which permit to create a virtual client with set number of outputs (2, 4, 6,8). The default will be stereo client, and this default will also connect its sound to physical outputs.

- The client will name itself using the setting of the application, or it falls back to `VERSION_NAME`, defined as "Godot Engine".

- JACK audio clients are expected to operate under hard-realtime conditions. While I don't know if `audio_server_process` will be, this driver code at least is expected to conform.

- The audio thread will try-lock the `Mutex`, not `lock`, this helps Godot to be a good client and not xrun the JACK server.

jpcima committed:
Audio driver for JACK, playback only

jpcima committed:
Add capture abilities to the JACK driver

jpcima committed:
Remove a comment which is not relevant [ci skip]

fire commented:
There is still interest in this, but are you able to update for master?

fire commented:
A concern about compile time dependency was mentioned that would block merging. 

I noticed that https://github.com/dr-soft/miniaudio/blob/master/miniaudio.h was able to activate JACK without the dependency.

jpcima committed:
Updates for Godot 4.0

jpcima committed:
Update the coding style

jpcima commented:
@fire, here's a rebase for now, I've updated but not yet completed a build.
I'm not sure yet of what's the trouble with miniaudio, I will have a look.

fire commented:
If it wasn't clear, the issue is that there's a compile time dependency on JACK.

miniaudio has done some work to avoid having the linker dependency.

jpcima committed:
Update the code style

fire commented:
There's still some style problems.

jpcima committed:
Load the JACK library at runtime

jpcima committed:
Allow the Unix platform to search libraries in system paths

jpcima commented:
Ok so this is an attempt. It's not necessary to copy from the miniaudio lib, I already did this job before many times.
I have some trouble to replicate the test process I did last time, but I'll try to figure it out with Godot 4.0.

fire commented:
After you are ready I can ping reduz for a code review, I believe hpvb is not able to review this promptly.

jpcima committed:
Adapt the python coding style

jpcima commented:
Ok I have got the latest working on my setup, so I'm fine to call it ready for a review.

fire commented:
Is it too difficult to get it to work on Windows? I have to setup a Linux machine for JACK.

Might take some time.

jpcima commented:
I have no idea, haven't worked on Windows for a long time.
I recall the Windows port is not well maintained, possibly for lack of interest.
There was a group of researchers interested to maintain a macOS version some time ago, not sure what the state of that is.
The Linux version is best without a doubt.

fire commented:
Since my Linux install Fedora 34 now uses Pipewire and it has a jack emulation library, I might this touching on this again.

I don't think Windows support is required.

jpcima committed:
Update to latest Godot

jpcima committed:
Style updates for clang-tidy

jpcima committed:
Revert "Allow the Unix platform to search libraries in system paths"

This reverts commit 89cebdf8d1c3cfbd5cb1926c21525bdbe48a6621.

jpcima committed:
Rewrite the libjack loader with dynload-wrapper

jpcima committed:
Make the GA workflow install libjack-dev

jpcima commented:
This has been a long time, sorry about that. I've been busy and this kind of flew over my head.
The work is rebased, updated to latest godot and code style.
Also, updated to dynload-wrapper, which is identical to what pulseaudio uses.

jpcima committed:
Fix some formatting problems

jpcima committed:
Remade libjack.so wrapper from Ubuntu 20.04 headers

jpcima committed:
Fix a python formatting problem

jpcima committed:
Install libjack-dev in the 2 other Linux workflows

jpcima committed:
Fix indexing the incorrect device

fire commented:
Consensus from the group was this is niche and there wasn't wide support. However, please don't delete the branch, I might find  a use for it in godot live performances.

Do you know know if pipewire has an alternative api?

jpcima commented:
> Do you know know if pipewire has an alternative api?

One could use the pipewire native API. There doesn't seem to be much benefit over using libjack directly however.
Using libjack would work equally on both JACK and Pipewire servers.

There is a relevant [FAQ item], which advises against using Pipewire APIs directly, at least for the time being.

[FAQ item]: https://gitlab.freedesktop.org/pipewire/pipewire/-/wikis/FAQ#what-audio-api-do-you-recommend-to-use

orowith2os commented:
Adding PipeWire and/or some other audio server support for Linux would be appreciated, as PulseAudio doesn't allow you to capture application-specific audio without some major hacks, and isn't very sandbox-friendly.

If JACK supports this functionality (specifically, capturing individual app audio, at least with pipewire-jack), I'd like to see this merged, or some discussions on it as of the current state of Linux and PipeWire support now.

The FAQ page on PipeWire seems to still suggest using pipewire-pulse, as it's one of the more maintained audio implementations.

WildRackoon commented:
This woud be very useful indeed for live performances, I wouldn't mind working on this.
Would it be more suited to maintain this in a separate godot module if people are reluctant to merge in core ?

starry-abyss commented:
> We mostly can't use miniaudio. So your changes can reference but not use it directly. However, leaving an attribution is good.

@fire I wonder why Godot can't use some miniaudio code directly? From MIT-0 license it seems to me that not only it's compatible with Godot's MIT, but actually doesn't even care if the attribution is given or not.

fire commented:
There has been discussion about implementing Pipewire for more composer—and artist-friendly audio routing, but I have not yet seen a pull request for It.
